js/src/ExpectedMark.js

```diff
@@ -0,0 +1,57 @@
+import UXCapture from "./UXCapture";
+
+const expectedMarks = [];
+const onMarkEventListener = null;
+
+/**
+ * Class describes expected marks
+ * These marks that have to be recorded before zone is considered complete
+ */
+export default class ExpectedMark {
+  static get(label) {
+    const mark = expectedMarks.find(mark => mark.label === label);
+
+    return mark;
+  }
+
+  static setOnMarkEventListener(callback) {
+    onMarkEventListener = callback;
+  }
+
+  constructor(markLabel) {
+    this.label = markLabel;
+    this.zones = [];
+
+    this.done = null;
+
+    this.promise = new Promise((resolve, reject) => {
+      this.done = () => {
+        UXCapture.recordMark(this);
```

Comment:
I think is one source of complexity because it's sort of circular - UXCapture tracks Zones, Zones track Marks, and Marks call UXCapture when they're 'done'. I think we could figure out a way to use callbacks passed to each class's constructor to keep each 'layer' independent. `UXCapture` could pass down its `config.onMeasure` and `config.onMark` callbacks to `Zone`, which would be responsible for calling `onMeasure` and passing `onMark` down to `Mark`, which would be responsible for calling `onMark`. That kind of setup would follow the React convention of passing down props to children and letting children use callbacks to talk to their parents rather than directly calling their parents' static methods.

Reply:
Yes, this is what I was leaning towards that, envisioned some code like this to create a zone object:
```javascript
new Zone(View)(marks)
```

Reply:
gotcha - unless a Zone needs to know its parent view, you could probably get away with something like

```js
class View {
  constructor(config) {
    const { zones } = config.zones;
    this.zones = zones.map(({ zone }) => new Zone({ marks }));
  }
}
```

where the parent view keeps track of Zones, but the Zone itself doesn't get any info about the view that owns it.